core/src/main/java/org/infinispan/container/DefaultDataContainer.java

```diff
@@ -100,30 +113,60 @@ protected DefaultDataContainer(int concurrencyLevel, long thresholdSize,
             throw new IllegalArgumentException("No such eviction thread policy " + strategy);
       }
 
-      Eviction eviction;
-      switch (strategy) {
-         case FIFO:
-         case UNORDERED:
-         case LRU:
-            eviction = Eviction.LRU;
+      Caffeine<K, InternalCacheEntry<K, V>> caffeine = caffeineBuilder();
+
+      switch (thresholdPolicy) {
+         case MEMORY:
+            CacheEntrySizeCalculator<K, V> calc = new CacheEntrySizeCalculator<>(new WrappedByteArraySizeCalculator<>(
+                  new MarshalledValueEntrySizeCalculator()));
+            caffeine.weigher((k, v) -> (int) calc.calculateSize(k, v)).maximumWeight(thresholdSize);
             break;
-         case LIRS:
-            eviction = Eviction.LIRS;
-            if (thresholdPolicy == EvictionType.MEMORY) {
-               throw new IllegalArgumentException("Memory based approximation eviction cannot be used with LIRS!");
-            }
+         case COUNT:
+            caffeine.maximumSize(thresholdSize);
             break;
          default:
-            throw new IllegalArgumentException("No such eviction strategy " + strategy);
-      }
-      EntrySizeCalculator<K, InternalCacheEntry<K, V>> sizeCalculator =
-            thresholdPolicy == EvictionType.MEMORY ? new CacheEntrySizeCalculator<>(
-                    new MarshalledValueEntrySizeCalculator()) : null;
+            throw new UnsupportedOperationException("Policy not supported: " + thresholdPolicy);
+      }
+      entries = new CaffeineConcurrentMap<>(applyListener(caffeine, evictionListener).build());
+   }
+
+   private Caffeine<K, InternalCacheEntry<K, V>> applyListener(Caffeine<K, InternalCacheEntry<K, V>> caffeine, DefaultEvictionListener listener) {
+      return caffeine.executor(new WithinThreadExecutor()).removalListener((k, v, c) -> {
```

Comment:
The writer is synchronous and performed in the `compute` block that is mutating the entry. So it will be run while the entry is still visible in the hash map. The removal listener is run after its been removed from the hash map, so it does not block map operations. If asynchronous then its preferred.

If you can use the default executor and the writer, then that's probably best. If the listener calls are expensive you might be able to hand-off synchronously and process asynchronously (e.g. via a write-behind buffer like in the RxJava example). I tried to keep costs well balanced so while a same-thread executor should be fine, I do prefer if possible hiding penalties from the caller.